flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/WaitingForResourcesTest.java

```diff
@@ -178,12 +344,24 @@ void setExpectCreatingExecutionGraph() {
             creatingExecutionGraphStateValidator.expectInput(none -> {});
         }
 
-        void runScheduledTasks() {
-            for (ScheduledTask<Void> scheduledTask : scheduledTasks) {
-                scheduledTask.execute();
+        void runScheduledTasks(long untilDelay) {
+            Iterable<ScheduledTask<Void>> copyOfScheduledTasks = new ArrayList<>(scheduledTasks);
+            Iterator<ScheduledTask<Void>> scheduledTaskIterator = copyOfScheduledTasks.iterator();
+            while (scheduledTaskIterator.hasNext()) {
+                ScheduledTask<Void> scheduledTask = scheduledTaskIterator.next();
+                if (scheduledTask.getDelay(TimeUnit.MILLISECONDS) <= untilDelay) {
+                    scheduledTask.execute();
+                    if (!scheduledTask.isPeriodic()) {
+                        scheduledTaskIterator.remove();
+                    }
+                }
```

Comment:
I think this implementation has the problem that we are forgetting to execute tasks which are created as part of other tasks because we copy `scheduledTasks`.

Reply:
But they will be executed with the next call to runScheduledTasks(). I want to only execute tasks which are scheduled up to now. 
Immediately executing newly scheduled tasks can lead to loops, for example WaitingForResources.checkDesiredOrSufficientResourcesAvailable() schedules itself in some circumstances.

But there was actually a flaw in my code: The removal mechanism doesn't work as intended: It removes elements from the copy. My intention was actually to execute non periodic tasks only once. I've addressed this now in the code.
If you disagree with any of these changes, I'm happy to address this in a hotfix alongside another pull request or a separate PR.

Reply:
`untilDelay` should define an upper bound for the tasks to execute if there is no tasks with a delay of `0` which enqueues itself over and over again.

The problem with this implementation is that this might simulate an interleaving which cannot happen with a proper executor.